diff --git a/keyboards/moonlander/keymaps/optozorax/arbitrary_keycode/include.h b/keyboards/moonlander/keymaps/optozorax/arbitrary_keycode/include.h
index 98ce165..63382ad 100644
--- a/keyboards/moonlander/keymaps/optozorax/arbitrary_keycode/include.h
+++ b/keyboards/moonlander/keymaps/optozorax/arbitrary_keycode/include.h
@@ -1,7 +1,5 @@
 #pragma once
 
-#error "Для начала прочитай https://github.com/klavarog/arbitrary_keycode и измени код QMK как там сказано, затем можно эту ошибку закомментировать."
-
 void press_arbitrary_keycode(uint16_t keycode, bool down) {
 	keyrecord_t record = {
 	  .event = {
diff --git a/quantum/keymap_common.c b/quantum/keymap_common.c
index 39ba4a0..b76c901 100644
--- a/quantum/keymap_common.c
+++ b/quantum/keymap_common.c
@@ -179,7 +179,11 @@ __attribute__((weak)) void action_function(keyrecord_t *record, uint8_t id, uint
 // translates key to keycode
 __attribute__((weak)) uint16_t keymap_key_to_keycode(uint8_t layer, keypos_t key) {
     // Read entire word (16bits)
-    return pgm_read_word(&keymaps[(layer)][(key.row)][(key.col)]);
+    if (key.use_custom_keycode) {
+        return key.custom_keycode;
+    } else {
+        return pgm_read_word(&keymaps[(layer)][(key.row)][(key.col)]);
+    }
 }
 
 // translates function id to action
diff --git a/quantum/rgb_matrix.c b/quantum/rgb_matrix.c
index ab9d4bd..1d6d6a7 100644
--- a/quantum/rgb_matrix.c
+++ b/quantum/rgb_matrix.c
@@ -186,6 +186,10 @@ void rgb_matrix_set_color(int index, uint8_t red, uint8_t green, uint8_t blue) {
 void rgb_matrix_set_color_all(uint8_t red, uint8_t green, uint8_t blue) { rgb_matrix_driver.set_color_all(red, green, blue); }
 
 bool process_rgb_matrix(uint16_t keycode, keyrecord_t *record) {
+    if (record->event.key.use_custom_keycode) {
+        return true;
+    }
+
 #if RGB_DISABLE_TIMEOUT > 0
     if (record->event.pressed) {
         rgb_anykey_timer = 0;
diff --git a/tmk_core/common/keyboard.h b/tmk_core/common/keyboard.h
index 711d24c..13d497f 100644
--- a/tmk_core/common/keyboard.h
+++ b/tmk_core/common/keyboard.h
@@ -27,5 +27,7 @@ extern "C" {
 typedef struct {
     uint8_t col;
     uint8_t row;
+    bool use_custom_keycode;
+    uint16_t custom_keycode;
 } keypos_t;
 
 /* key event */
